src/coreclr/jit/gentree.cpp

```diff
@@ -14411,44 +14412,46 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
                 }
                 else
                 {
-                    // The tree under the box must be side effect free
-                    // since we will drop it if we optimize.
-                    assert(!gtTreeHasSideEffects(op->AsBox()->BoxOp(), GTF_SIDE_EFFECT));
+                    bool wrapEffects = true;
+                    if (op->IsBoxedValue())
+                    {
+                        // The tree under the box must be side effect free
+                        // since we will drop it if we optimize.
+                        assert(!gtTreeHasSideEffects(op->AsBox()->BoxOp(), GTF_SIDE_EFFECT));
 
-                    // See if we can optimize away the box and related statements.
-                    GenTree* boxSourceTree = gtTryRemoveBoxUpstreamEffects(op);
-                    bool     didOptimize   = (boxSourceTree != nullptr);
+                        // See if we can optimize away the box and related statements.
+                        wrapEffects = (gtTryRemoveBoxUpstreamEffects(op) == nullptr);
```

Comment:
previous logic used to give up if box can't be removed, is it expected that the new one always folds?

Reply:
btw, does it all handle boxed nullables (when boxed value is null reference)?

Reply:
> is it expected that the new one always folds?

Yeah, my reasoning here was that there's no reason to avoid folding if we can't remove.

> does it all handle boxed nullables (when boxed value is null reference)?

Not sure, I'd assume those wouldn't pass `IsBoxedValue`, otherwise I'd think the previous code would be broken.